updated threatmodel using supervisor domains terms (#35)
riscv-non-isa/riscv-ap-tee: specification/threatmodel.adoc

T19: Stale TLB translations (for U/HS mode or for VU/VS) created during TSM
or TVM operations are used to execute malicious code in the TVM (or consume
stale/invalid data)

T20: Isolation of performance monitoring and/or debug state for a TVM
leading to information loss via performance monitoring events/counters and
debug mode accessible information.

T21: A TVM causes a denial of service on the platform

[NOTE]
====
This threat model is not an exhaustive list and will be updated (via the RISC-V
Security Model specification <<R21>>) on a regular basis as attacks evolve.
====

=== Scope

This specification does not prescribe the scope of mitigation and focusses
on the TEEI interface and use-of/impact-on the RISC-V ISA. It is
recommended that implementations of this reference architecture address
threats from system software adversaries. Implementations may choose to
mitigate threats from additional adversaries. For all cases, denial of
service by TVMs must be prevented. At the same time, denial of service by
host software is considered out of scope.

[[design_survey]]
=== TVM Security Requirements to address threat model

> 

Reviewer: All these are just space removal changes until this point ?

Author: yes - no major changes in this section :)

[[threatmodel]]
=== Adversary Model

_Unprivileged Software adversary -_ This includes software executing in
U-mode managed by S/HS/M-mode system software. This adversary can access
U-mode CSRs, process/task memory, CPU registers in the process context

> managed by system software.

Reviewer 2: With user space I/O an Unprivileged adversary may also have ability to submit requests to I/O devices.

Author: added this vector.

> _Unprivileged Software adversary -_ This includes software executing in
> U-mode managed by S/HS/M-mode system software. This adversary can access
> U-mode CSRs, process/task memory, CPU registers in the process context
> managed by system software.

Author: ```suggestion
_Unprivileged Software adversary -_ This includes software executing in
U-mode managed by S/HS/M-mode system software. This adversary can access
U-mode CSRs, process/task memory, CPU registers in the process context
managed by system software. With user space I/O an Unprivileged software
adversary may also have ability to submit requests to I/O devices made 
available by system software for U-mode access.
```

> _System Software adversary_ - This includes system software executing in
> S/HS/VS modes. Such an adversary can access S/HS/VS privileged CSRs,
> assigned system memory, CPU registers and IO devices.

Author: ```suggestion
_System Software adversary_ - This includes system software executing in
S/HS/VS modes. Such an adversary can access S/HS/VS privileged CSRs,
assigned system memory, CPU registers, IOMMU(s) and IO devices.
```

> 

Reviewer 2: and IOMMU.

Author: ack

> _Startup Software adversary_ - This includes system software executing in
> early/boot phases of the system (in M-mode), including BIOS, memory
> configuration code, device option ROM/firmware that can access system
> memory, CPU registers, IO devices and IOMMU etc.

Author: ```suggestion
_Startup Software adversary_ - This includes system software executing in
early/boot phases of the system (in M-mode), including BIOS, memory
configuration code, device option ROM/firmware that can access system
memory, CPU registers, IOMMU(s), IO devices and platform configuration 
registers (e.g., address range decoders, SoC fabric configuration, etc.).
```

> 

Reviewer 2: and platform configuration registers (e.g., address range decoders, SoC fabric configuration, etc.).

Author: ack

> _Non-invasive Hardware adversary_ __ - This includes adversaries that can use
> non-invasive hardware attacks such as bus interposers to snoop on memory/device
> interfaces, voltage/clock glitching, observe electromagnetic and other
> radiation, analyze power usage through instrumentation/tapping of power
> rails, etc. which may give the adversary the ability to tamper with data in
> memory.

Author: ```suggestion
_Non-invasive Hardware adversary_ - This includes adversaries that can use
non-invasive (requiring no physical change to the target hardware) attacks 
such as bus interposers to snoop on memory and/or device interfaces, 
voltage and/or clock glitching, observe electromagnetic and other radiation, 
analyze power usage through instrumentation/tapping of power rails, etc. 
which may then give the adversary the ability to tamper with data in use.
```

> 

Reviewer 2: How would non-invasive be defined? One could interpret adding instrumentation or tapping signals as being invasive.

Author: non-invasive attacks require no physical change to the target hardware

other shared resources e.g. via prime/probe, flush/reload approaches

T16: Malicious access to ciphertext with known plaintext to launch a
dictionary attack on TCB components to extract confidential data.

T17: Tamper of TVM state during migration of a TEE workload assets within
the platform or from one platform to another.

T18: Forging of attestation evidence and sealed data associated with a TVM.

T19: Stale TLB translations (for U/HS mode or for VU/VS) created during TSM
or TVM operations are used to execute non-TCB code in the TVM (or consume
stale/invalid data)

T20: Isolation of performance monitoring and/or debug state for a TVM
leading to information loss via performance monitoring events/counters and
debug mode accessible information.

T21: A TVM causes a denial of service on the platform

[NOTE]
====
This threat model is not an exhaustive list and will be updated (via the RISC-V
Security Model specification <<R21>>) on a regular basis as attacks evolve.
====

=== Scope

This specification describes the threats that a system implementing CoVE should
address, however, it does not prescribe the scope of mitigations; instead it

> focusses on mitigations enabled via the COVH/G interface via the underlying use
> of the RISC-V ISA (and extensions such as Smmtt). This specification also

Author: ```suggestion
focusses on mitigations enabled via the COVH/G interface and the use
of the RISC-V ISA (and extensions such as Smmtt). This specification also
```

service by non-TCB software (e.g. in a hosting supervisor domain) is considered
out of scope.

[[design_survey]]
=== TVM Security Requirements to address threat model

|===
| *Category*  |  *Security Criteria*  |  *CoVE Requirement* |  *Example methods
 to meet requirement* | *Description / Example* | *RVI normative or
non-normative Reference*

| Memory Assignment | Ability to make memory confidential or non-confidential |
Required | MMU, MPU, MTT | Confidential memory should be dynamically allocated/
unallocated as required | RISC-V Priv. ISA, Smmtt

| TEE CPU State Protection | State Isolation | Required | Priv. levels (M,S,HS,
U) and Execution context (ASID, VMID, SDID) | Prevent non-TCB components from
arbitrarily accessing/modifying TEE CPU state | Priv ISA, Sv, Smmtt

| Memory Confidentiality | Memory isolation (read) | Required | cryptography
 and/or MMU, MPU, MTT extension | Prevent non-TCB components from reading
TEE memory | Sv, Smmtt

| Memory Confidentiality | Cipher text read prevention | Required | cryptography
 and/or MMU, MPU, MTT extension | Prevent non-TCB components from accessing
encrypted TEE memory | Svpams (Supervisor Domains)

> | Memory Confidentiality | Per TEE encryption | Implementation-specific |
> cryptography and/or MMU, MPU, MTT extension | Each VM has one or more
> unique keys  | Svpams (Supervisor Domains)

Author: ```suggestion
| Memory Confidentiality | TEE encryption | Implementation-specific |
Cryptography and/or MMU, MPU, MTT extension | TEE memory confidentiality 
against logical attacks via MMU; additionally address physical attacks via 
cryptography. If cryptography used, TEE should have a unique encryption key; 
each TEE VM may may also have one or more unique keys. Also see related 
requirements around ciphertext disclosure and memory integrity  | Svpams 
(Supervisor Domains)
```

> 

Reviewer 2: Why is MMU/MPU/MTT not sufficient to provide memory confidentiality property? How does per-TEE encryption provide confidentity. Per TEE eencryption would still not allow letting one TEE access another TEEs ciphertext due to the ciphertext disclosure issue. By corollary if the memory region is not accessible in first place then fact that it is per-TEE encrypted or not may be moot?

Author: hence the "or" - this is against physical attacks - agree that it goes hand-in-hand with preventing ciphertext disclosure, as well as memory integrity and replay-protection. Adding a note to the effect (ptal at updated text).

and/or MMU, MPU, MTT | Prevent TEE from executing from normal memory; Enforce
integrity of TEE data on writes | CoVE ABI

| Memory Integrity | Rowhammer attack prevention | Implementation-specific |
cryptography and/or memory-specific extension | Prevent non-TCB components from
flipping bits of TEE memory | Security Model

| Shared Memory | TEE controls data shared with non-TCB components | Required |
cryptography and/or MMU, MPU, MTT | Prevent non-TCB code from exfiltrating
information without TEE consent/opt-in | CoVE ABI

| Shared Memory | TEE controls data shared with another TEE | Implementation
specific | cryptography and/or MMU, MPU, MTT | Ability to securely share memory
with another TEE | CoVE ABI, Svpams

| I/O Protection | DMA protection from non-TCB-admitted devices | Required | DMA
access-control e.g. IOPMP, IOMTT, IOMMU | Prevent non-TCB peripheral devices
from accessing TEE memory | CoVE-IO, IOMMU, IOMTT

| I/O Protection | Trusted I/O from devices admitted into the TCB of a TVM |
Implementation-specific | Device attestation, Link protection, IOMMU |
Admission control to bind devices to TEEs | CoVE-IO, IOMMU

> | Secure IRQ | Trusted Interrupts | Required | Secure interrupt files, MMU, MPU,

Author: ```suggestion
| Interrupts | Trusted (no spoofing/tampering/dropped) Interrupts | Required | Secure interrupt files, MMU, MPU,
```

 MTT | Prevent IRQ injections that violate priority or masking | Smsdia w/ AIA

> 

Reviewer 2: IRQ is a software term. IRQ->Interrupt.

Author: ack

| Secure Timetamp | Trusted timestamps | Required | Confidential supervisor
domain qualifier for CSR accesses | Ensure TEE have consistent timestamp view |
CoVE ABI

| Debug & Profile | Trusted performance monitoring unit data | Required |
Confidential supervisor domain context switch of perf. mon. counters  |
Ensure TEEs get correct PMU info; prevent data leakage due to PMU information
(fingerprint attacks) | CoVE ABI

| Debug & Profile | Secure External Debug support | Required | Confidential
supervisor domain qualifier for External debug controls | Support debug trigger
registers for TVM | Secure Debug, CoVE ABI

| Debug & Profile | Authenticated debug (Production device) | Required |
Authorize debug via TEE RoT | Ensure hardware debug prob (e.g., JTAG SWD) is
disabled in production | Secure Debug, CoVE ABI

| Availability | TVM DoS Protection | Required | VMM retains ability to
interrupt TVM | Prevent TVM from refusing to exit | CoVE ABI, Smsdia

| Availability | VMM DoS Protection | Implementation-specific | Not in scope
for CoVE | Prevent non-TCB components from refusing to run TEE | Not applicable

> | Side Channel | Protected address mapping (controlled side channel) | Required
> | Confidential mode qualifier, cryptography, MMU/MPU, MTT | Similar to memory
> remapping attacks | CoVE ABI, Security Model

Author: ```suggestion
| Side Channel | Address mapping caches (controlled side channel) | Required
| Confidential mode qualifier, MMU/MPU, MTT | HW/SW TCB should use 
tagging/ partitioning/ flushing techniques to address those types of side 
channels due to temporal/spatial shared resources | CoVE ABI, Security Model
```

> 

Reviewer 2: This is not clear. Protected address mappings do not defend against classical set-collision type cache based side channels.

Author: this was referring to TLBleed like attacks - updated to specify TCB should use tagging/partitioning/flushing techniques to address those types of side channels.

> | Side Channel | Micro-architectural side channels (branch prediction | Required
> | uArch state flushing, entropy defenses | Prevent attacks such as meltdown or
> spectre (it is difficult to defend against such attacks in advance) | Security
> Model

Author: ```suggestion
| Side Channel | Transient-execution attack (TEA) side channels | Implementation-specific
| * Bounds check bypass TEA and variants - should be addressed by TVM software 
using apropos synchronization. TCB SW should use synchronization to isolate TCB code 
from non-TCB code.
* Branch target injection TEA and variants - should be addressed by TCB SW via 
flushing across privilege boundaries to remove untrusted state injected by 
non-TCB software
* Speculative store bypass TEA and variants - should be addressed by TCB HW 
via synchronization/barriers to prevent speculative execution of memory reads 
which may allow unauthorized disclosure of information.
| Implementations should mitigate attacks such as these spectre variants (In 
practice, it is difficult to defend against such attacks in advance) | 
Recommendations in Security Model
```

> 

Reviewer 2: Seems to mix microarchitectural side channel and transient execution techniques. Two should be seprated.

Author: enumerating the sub-types in this one - ptal.

| Side Channel | Control channels, single-step/zero-step attacks | Required |

> uArch state flushing, entropy defenses | Prevent interrupt/exception injection

Author: ```suggestion
leverage HW/SW TCB mechanisms to enforce restrictions on single-stepping 
or zero-stepping via use of state flushing/barriers, entropy defenses and 
detection mechanisms. | Prevent interrupt/exception injection
```

(combined with cache side channel to leak sensitive data) | Security Model

> 

Reviewer 2: Unclear what uarch state flushing and entropy defences imply.

Author: leverage via HW/SW TCB mechanisms to enforce restrictions on single-stepping or zero-stepping via use of state flushing/barriers, entropy defenses and detection mechanisms.

| Side Channel | Architectural cache side channel | Implementation-specific  |

> uArch state flushing, entropy defenses | Prevent shared resource contention,

Author: ```suggestion
cache partitioning-based defenses | Prevent shared resource contention,
```

e.g. attacks such as prime probe | Security Model

> 

Reviewer 2: Not clear what/how uarch state flushing or entropy defenses help against cache side channels.

Author: yeah this was a copy/paste error - will mention cache partitioning as an example strategy for this case.

| Side Channel | Architectural timing side channel | Implementation-specific  |
data independent execution latency (DIEL) operations, uArch state flushing |
Leveraging data dependency timing channels | Security Model

| Secure and measured boot | Establishes root of trust in support of attestation
| Required | RoT unique trust chain for TEE TCB | Enforcing initial firmware
authorization and versioning | Security Model

> | Attestation | Remote attestation | Required | HW RoT based PKI (trust

Author: ```suggestion
| Attestation | Remote attestation | Required | HW-RoT-rooted PKI (trust
```

assertions) via Internet | Prevent fake hardware and software TCB; Prevent
non-TCB hardware debugging in production. | CoVE ABI, Security Model

> 

Reviewer 2: based or rooted?

Author: rooted

| Attestation | Mutual attestation | Implementation-specific | S/U mode  |
Attestation to another TEE on the same platform | CoVE ABI

> | Attestation | Remote mutual attestation | Required | Internet |  Attestation
> to a relying party on a different platform | CoVE ABI

Author: ```suggestion
| Attestation | Remote mutual attestation | Required | Internet |  Attestation
to a relying party on a different platform. Requires provisioning of the TEEs 
to act as delegated relying parties | CoVE ABI
```

> 

Reviewer 2: how different than remote attestation?

Author: requires additional provisioning of the TEEs to act as relying parties. agree its a nuance - will add a clarification for now - if its confusing, can remove it.

| Attestation | Local attestation | Implementation-specific | Sealing |
Verification of attestation by TCB | Future CoVE ABI

> | Attestation | TCB versioning (and updates) | Required | Mutable firmware where
>  TVM has to opt-in if TCB updates are allowed or not - HW TCB then enforces
> lower TCB elements are updatable (with apropos controls like SVN) only after
> that opt-in has been honored. | Allow TCB updates - Prevent TCB rollback |
> CoVE ABI

Author: ```suggestion
| Attestation | TCB versioning (and updates) | Required | Mutable firmware where
 TVM has to opt-in at startup if TCB updates are allowed while the TVM is 
 executing - HW TCB then enforces lower TCB elements are updatable 
 (with apropos controls like security version enforced) to enforce the 
 opt-in policy. | Allow TCB updates - Prevent TCB rollback | CoVE ABI
```

> 

Reviewer 2: The opt-in should be to execute across a TCB update. If not opted in the TVM terminates.

Author: updated description to clarify that.
